feat(postgres): add name option for enum type

sturdynut opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Closes #2577

These changes allow for an optional name when defining enums.

### Notes

* I made a change in the `test/integration/model/count.test.js` file.  This was outside of the scope of the feature in I was working on, it was the only linter error and thought the change was pretty minimal enough to slide in.
* In one of the new enum name tests in `.../postgres/query-interface.test.js`, I ended up using the `done()` callback _(which I know is called out as something not to do.)_ , in order to get an accurate failure message.  Otherwise the test just times out and you don't see the error from the failed `expect(...)...`;
* In `.../postgres/data-types.js` I wasn't a huge fan of the `type.type || type` stuff on line 480, but I just don't understand the code base enough to make an informed refactoring here.  Maybe it's fine, but just something I noticed as I was making changes.
* In `.../postgres/query-generator.js` I updated the signature for the `pgEnumDrop(...)` method and reluctantly settled on renaming the `enumName` parameter to `dataTypeOrEnumName`. I did this in order to not break existing calls that provided the name where the dataType was not available.  Otherwise I'd have just renamed it to `dataType` and passed that in for all cases.
* When running the integration tests, I received 3 failures prior to making any changes.  Just wanted to make sure this was known and that I didn't in fact cause it, which seems unlikely.

These are the existing broken integration tests:

```
  1) [POSTGRES Specific] Data Types
    DATEONLY SQL
      should be able to create and update records with Infinity/-Infinity:
  AssertionError: expected Sat Oct 05 2019 to be within Sun Oct 06 2019 and Sun Oct 06 2019

  2) [POSTGRES Specific] ExclusionConstraintError
    "before each" hook for "should contain error specific properties":
  SequelizeUnknownConstraintError: Unknown constraint error

  3) [POSTGRES Specific] QueryInterface
    createFunction
      uses declared variables:
  SequelizeDatabaseError: function "add_one" already exists with same argument types
```

kwakwaversal commented on test/unit/sql/enum.test.js:
```diff
@@ -24,6 +24,26 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
         }
       });
 
+      const NamedEnumUser = current.define('user', {
+        mood: {
+          type: DataTypes.ENUM({
+            name: 'mood',
+            values: ['happy', 'sad']
+          })
+        }
+      });
```
Should there be a test case for an `ARRAY` of named `ENUM`s?

sturdynut replied:
Happy to add a test, but I don't quite follow.  Can you provide an example of exactly what you mean?

kwakwaversal replied:
Sorry for the late reply. Something like:

```
const NamedEnumUser = current.define('user', {
  mood: {
    type: DataTypes.ARRAY(DataTypes.ENUM({
      name: 'mood',
      values: ['happy', 'sad']
    }))
  }
});
```

To ensure an `ARRAY`s of named enums works.

sturdynut replied:
Thank you, I will test for this.

sturdynut replied:
I wrote the test for this and realized how this still does not seem right.  I'm not sure of a case where you'd want an array of enums for the type.  Apologies, maybe this is over my head but as far as I know, when you have a column in PG that is of type enum...there is only one enum type that it is tied to.  Can you help me understand a bit more?

nmchaves replied:
I think they're talking about a scenario like this:
```sql
CREATE TABLE person (
    name text,
    current_moods mood[]
);
```

...where the `mood[]` type allows the person to have multiple moods simultaneously.

sturdynut replied:
Ahh, right.  That makes sense.  Thank you.

nmchaves replied:
Thank _you_ for working on this PR! 👏

kwakwaversal replied:
As @nmchaves said.

An enumerated type in PostgreSQL is typically a `text` type. So think of it as me suggesting an `array` of `text` which is perfectly valid. The only difference by using an enum instead of `text` is obviously that the array values can only be one of the defined constants.

papb commented on test/integration/model/count.test.js:
```diff
@@ -53,7 +53,7 @@ describe(Support.getTestDialectTeaser('Model'), () => {
         {
           username: 'valak',
-          createdAt: (new Date()).setFullYear(2015)
+          createdAt: new Date().setFullYear(2015)
```
> * I made a change in the `test/integration/model/count.test.js` file.  This was outside of the scope of the feature in I was working on, it was the only linter error and thought the change was pretty minimal enough to slide in.

No problem!

papb commented on test/integration/dialects/postgres/query-interface.test.js:
```diff
@@ -287,5 +287,38 @@ if (dialect.match(/^postgres/)) {
+            done();
+          })
+          .catch(done);
+      });
```
> * In one of the new enum name tests in `.../postgres/query-interface.test.js`, I ended up using the `done()` callback _(which I know is called out as something not to do.)_ , in order to get an accurate failure message.  Otherwise the test just times out and you don't see the error from the failed `expect(...)...`;

Hmmm, I doubt it is really necessary, I suspect you did something wrong, can you please try the code below instead?

```js
it('should work with named enums', function() {
    return this.queryInterface.addColumn('users', 'mood', {
        type: DataTypes.ENUM({
            name: 'mood_enum',
            values: ['happy', 'sad', 'neutral']
        })
    }).then(() => this.sequelize.query(
        'SELECT udt_name FROM information_schema.columns WHERE table_name = \'users\' AND column_name = \'mood\'',
        { type: this.sequelize.QueryTypes.SELECT }
    )).then(res => {
        expect(res[0].udt_name).to.be.equal('mood_enum');
    });
});
```

sturdynut replied:
I see, I forgot to return my promise; that works.  Thank you.

papb commented on lib/dialects/postgres/data-types.js:
```diff
@@ -477,10 +477,11 @@ module.exports = BaseTypes => {
-          Utils.generateEnumName(options.field.Model.getTableName(), options.field.fieldName),
-          '"'
-        ) }[]`;
+        const enumType = this.type.type || this.type;
```
> * In `.../postgres/data-types.js` I wasn't a huge fan of the `type.type || type` stuff on line 480, but I just don't understand the code base enough to make an informed refactoring here.  Maybe it's fine, but just something I noticed as I was making changes.

Weird indeed, but are you sure this is necessary? What happens when using the code below instead?

```suggestion
        const enumType = this.type;
```

sturdynut replied:
Made the change, re-ran tests and still passed, so should be good!

Here are the other places where I've added this same check.  Tests failed when I removed either of these, so assume they are ok.

File: `/lib/dialects/postgres/query-generator.js`
Line 466: `const enumType = attribute.type.type || attribute.type;`
Line 764: `const type = dataType.type || dataType;`

Not sure if there is any refactoring that should be done here, but just wanted to make note that the reason I originally added these checks were because I saw the code below:

File: `/lib/dialects/postgres/query-interface.js`
Line 84: `const enumType = type.type || type;`

papb commented:
Hello! I see you are a first-time contributor, thank you for taking the time to help Sequelize! I hope to see more PRs from you in the future!

I have made a first review to your PR above. Thanks for your work :)

> * In `.../postgres/query-generator.js` I updated the signature for the `pgEnumDrop(...)` method and reluctantly settled on renaming the `enumName` parameter to `dataTypeOrEnumName`. I did this in order to not break existing calls that provided the name where the dataType was not available.  Otherwise I'd have just renamed it to `dataType` and passed that in for all cases.

This part I haven't looked yet, will do later.

> * When running the integration tests, I received 3 failures prior to making any changes.  Just wanted to make sure this was known and that I didn't in fact cause it, which seems unlikely.
> 
> These are the existing broken integration tests:
> 
> ```
>   1) [POSTGRES Specific] Data Types
>     DATEONLY SQL
>       should be able to create and update records with Infinity/-Infinity:
>   AssertionError: expected Sat Oct 05 2019 to be within Sun Oct 06 2019 and Sun Oct 06 2019
> 
>   2) [POSTGRES Specific] ExclusionConstraintError
>     "before each" hook for "should contain error specific properties":
>   SequelizeUnknownConstraintError: Unknown constraint error
> 
>   3) [POSTGRES Specific] QueryInterface
>     createFunction
>       uses declared variables:
>   SequelizeDatabaseError: function "add_one" already exists with same argument types
> ```

Looks like this only happened locally for you, since all checks have passed in the PR, so don't worry about it. It must be something about your local postgres setup.

![image](https://user-images.githubusercontent.com/20914054/66346210-0d24f400-e928-11e9-897b-edfe3041e1ab.png)

papb commented:
Closes #2577

sturdynut commented:
@papb - Thank you!  I've used sequelize on many projects and it is a pleasure to help out. 😃

I've pushed my latest updates based on your feedback.
  * Returned promise and removed `done()` callback
  * Removed unnecessary `.type` check

papb commented:
Sorry I don't have time to look into it right now... Will do later, hopefully, or maybe another maintainer will show up :)

sturdynut commented:
@papb - All good!  I just wanted to make sure you knew that it was ready for re-review.   Thank you!

sushantdhiman commented:
Please remove unrelated name changes etc from this branch

sturdynut commented:
@sushantdhiman -Are you referring to where I've renamed a parameter because I added another? 
 An example is where I've renamed "attributes" to "sqlAttributes".  I'm happy to make the change I just want to be sure I fully understand.

![image](https://user-images.githubusercontent.com/8547391/68437464-df74da00-0175-11ea-8a85-f27f9119f37f.png)

papb commented:
Hello @sturdynut, thanks for the patience so far. I am not sure what @sushantdhiman was referring to... Perhaps the change from `dataType` to `dbDataType`? That's fine by me though...

sturdynut commented:
@papb - No problem.  I'm happy to make whatever naming changes necessary.  Once @sushantdhiman confirms direction I will proceed.

papb reviewed (changes requested):
How does this PR relate to `sync({ alter: true })` (#7649)?

I think it would be great if this PR could handle both issues at the same time (and if luckily it already fixes `alter: true` as is, please add a test for that).

sturdynut commented:
@papb - This certainly is in the same code.  I can add the check for existing enum mentioned in https://github.com/sequelize/sequelize/issues/7649

Still waiting on confirmation as to how to proceed on this PR.  I can revert any name changes I made, but I just want to confirm before I slate more time to work on this.

Is it just the name change from `dataType` to `dbDataType`? @sushantdhiman

papb commented:
Thanks @sturdynut, let's wait for @sushantdhiman then :)

ShahoG commented:
Any news on this? Much appreciated feature

sturdynut commented:
@ShahoG - I was waiting to hear back from @sushantdhiman before moving forward.  If I don't hear anything by this weekend, i'm going to make the changes I believe were requested along with the fix from #7649 

I don't like having this sitting too long.  Cringing thinking about potential merge conflicts, but will make traction on this over the weekend regardless.

sushantdhiman committed:
refactor: use compiled src from typescript (#12437)

sushantdhiman committed:
docs: note about master

xin-hao-2025 committed:
docs(ENGINE.md): remove beta (#12442)

sushantdhiman committed:
docs(readme): release documentation

kf6kjg committed:
feat(query-interface): add support for defining composite foreign keys (#12456)

patrickcarnahan committed:
fix(mssql): upsert query with falsey values (#12453)

sushantdhiman committed:
refactor: convert config maps & utils to typescript (#12464)

sushantdhiman committed:
refactor: use enum for suitable options

sushantdhiman committed:
refactor: convert errors to typescript (#12471)

tpodom committed:
fix: use lodash _baseIsNative directly instead of _.isNative (#12358) (#12475)

knoid committed:
fix(upsert): set isNewRecord to false when upserting (#12447)

renovate[bot] committed:
build: configure renovate (#12476)

renovate[bot] committed:
chore(deps): pin dependencies (#12490)

sushantdhiman committed:
build: renovate maintain lockfile

renovate[bot] committed:
chore(deps): update mariadb docker tag to v10.5 (#12491)

sushantdhiman committed:
build: rename mariadb container

DRoet committed:
fix: mark database drivers as optional peer dependencies (#12494)

eseliger committed:
feat(ts): add bundled types for toposort-class (#12497)

eseliger committed:
Migrate ModelManager to TypeScript (#12498)

sushantdhiman committed:
docs: remove myself from CONTACT.md

renovate[bot] committed:
chore(deps): update dependency chai-datetime to v1.7.0 (#12493)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

papb committed:
docs: add myself to CONTACT.md (#12503)

eseliger committed:
feat(ts): convert sql-string to ts (#12500)

* feat(ts): convert sql-string to ts

* fixup! feat(ts): convert sql-string to ts

Co-authored-by: Simon Schick <demwizzy@gmail.com>

JuarezLustosa committed:
docs(contributing): Explain to `npm run tsc` (#12510)

renovate[bot] committed:
fix(deps): update dependency lodash to v4.17.19 [security] (#12525)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

cahyonobagus committed:
docs(naming-strategies): fix user model name user (#12576)

mvp-studio-repo committed:
docs(migrations): add example for conditional unique index (#12578)

Co-authored-by: Valentin Almeida--Lemaire <valentin.almeida@onrewind.com>
Co-authored-by: Pedro Augusto de Paula Barbosa <papb1996@gmail.com>

ZeroCho committed:
docs(advanced-many-to-many): add missing `await` (#12533)

fernandomaia committed:
docs(raw-queries): update broken link (#12542)

vishal-sood committed:
fix(model): handle `true` timestamp fields correctly (#12580)

Co-authored-by: Pedro Augusto de Paula Barbosa <papb1996@gmail.com>

papb committed:
chore(deps): update some dependencies

renovate[bot] committed:
chore(deps): update dependency delay to v4.4.0 (#12553)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

papb committed:
Stop renovate trying to pin dependencies

iwaduarte commented:
any updates on this?  please @sturdynut save us.

sturdynut commented:
@papb - I had to do a hefty rebase and re-worked some of my initial work, in hopes that it addresses the renames that @sushantdhiman had mentioned.  I also added additional testing that was mentioned in the comments.  Looks like there is an open PR for #7649 (https://github.com/sequelize/sequelize/pull/12019), so I will not be doing any work on that here.

Will keep an eye out for any feedback or requested changes.

Commit:
feat(postgres): add name option for enum type

sturdynut commented:
@kwakwaversal / @papb - Any chance you could re-review?  🙏   I just don't want this to fall behind master again.  It took me a while to re-base and I know folks would love to have this feature!

iwaduarte commented:
@papb  How about that ?

sturdynut commented:
Just checking in here...heard this was a feature that some folks wanted...

kwakwaversal commented:
Looks good to me @sturdynut but I'm just a dev that required the feature you were working on and noticed there wasn't a test for arrays of specific ENUM types.

Not sure who to summon in the core dev team to bring attention to this if it's not @papb.

iwaduarte commented:
Not yet ?

ruigazio commented:
I would sure like to have the feature merged @papb

edivanteixeira commented:
This feature is very useful, i like to see this merged.

laurelgr commented:
This feature would be easier than having to use litteral queries in migrations or extend DataTypes.

How would you use it beyond adding the name? Does it create the enum if it doesn't exist, or should the enum already exists in postgres?
What if the values provided do not match the existing type, or each other across multiple models?

sturdynut commented:
> This feature would be easier than having to use litteral queries in migrations or extend DataTypes.
> 
> How would you use it beyond adding the name? Does it create the enum if it doesn't exist, or should the enum already exists in postgres?
> What if the values provided do not match the existing type, or each other across multiple models?

This feature simply allows you to provide a name for an ENUM type vs having sequelize create it for you using the table name and field name.   For existing enums, it already is named, so it wouldn't really help with changing the name, you'd likely need to drop your enum and re-create it.

Example:
```
DataTypes.ENUM({
    name: 'my_enum',
    values: ['value', 'another value']
 })
 ```

svanda-navisys commented:
I'm sorry to bother everyone, I would love to see this merged, is there anything specific that is holding this PR back?

angi- commented:
@sturdynut I'm looking forward to this PR!

How would I reuse the enum, something like this?
```
const statusEnum = DataTypes.ENUM({
    name: 'statuses',
    values: ['a', 'b', 'c']
});

FilterData.init({
    baseStatus: statusEnum,
    containsStatuses: DataTypes.ARRAY(statusEnum),
    notContainsStatuses: DataTypes.ARRAY(statusEnum)
});
```

How would I share this enum across multiple model files and migrations?

ephys commented:
This is something I've wanted for a while so I'll likely review it as some point. Can't promise when, work is keeping me busy again :/

sturdynut commented:
@ephys - It's been awhile since I looked at this myself, but will keep an eye out for any feedback you provide. 🙌🏼

ephys commented:
Just noticed this PR is targeting `master`. We renamed the branch to `main` a little while ago. It's likely there are going to be a few merge conflicts

I can handle it if you don't have time but it will be after the [PRs I'm currently working on](https://github.com/sequelize/sequelize/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3Aephys)

sturdynut commented:
> Just noticed this PR is targeting `master`. We renamed the branch to `main` a little while ago. It's likely there are going to be a few merge conflicts
> 
> I can handle it if you don't have time but it will be after the [PRs I'm currently working on](https://github.com/sequelize/sequelize/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3Aephys)

Just updated to `main`.  Yup, definitely some conflicts!  I don't have time at the moment to resolve, I will check back in here if I free up and can start to tackle them myself so we don't duplicate efforts.

ephys commented:
We are refactoring DataTypes in https://github.com/sequelize/sequelize/pull/14505 and, to reduce special-casing, the ENUM DataType is now responsible for generating its SQL name.

As such, adding a 'name' option was trivial and has been included in that PR.

iwaduarte commented:
I know you have a lot on your plate right now. What would we need to have that merged in v5 and v6? This is a 3-year-old pull request and although refactoring DataTypes would help, shouldn't we resolve this first?

ephys commented:
Sequelize 5 has been EOL for almost a year now.
We're not opposed to merging this in v6 but as you said, this PR is 3 years old. We have more active maintainers now but there are so many conflicts in this PR we may as well start over.

I personally focus more on fixing the foundations, but I'll review an updated PR that does this in v6 (but not v7 until https://github.com/sequelize/sequelize/pull/14505 has been reviewed and merged)
